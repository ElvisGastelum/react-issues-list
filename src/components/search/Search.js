import React from 'react';

import SearchItemsList from '../search-items-list';

import './index.css';

export default function Search({ newSearchResults, searchResults, index, setArrowKeys }) {

  const handleText = e => newSearchResults(e.target.value)
  const handleArrowKeys = e => setArrowKeys(e.keyCode);
  return (
    <div className="search-container">
      <div className="search-box">
        <input
          type="text"
          placeholder="Search"
          className="search-input"
          onChange={ handleText }
          onKeyDown= { handleArrowKeys }
        />
      <SearchItemsList items={ searchResults } index={ index } setArrowKeys={ setArrowKeys } />
      </div>
    </div>
  )
}
