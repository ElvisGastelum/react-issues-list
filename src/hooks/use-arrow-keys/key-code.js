export const key = {
  UP: 38,
  DOWN: 40,
  ENTER: 13,
  MOUSE_EVENT: -1,
}
