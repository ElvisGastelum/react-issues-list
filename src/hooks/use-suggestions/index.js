import { useState } from 'react';

import { useGithubIssuesList } from '../use-github-issues-list'

export const useSuggestions = (initialState) => {
  const list = useGithubIssuesList([]);
  const [searchResults, setSearchResults] = useState(initialState);



  const newSearchResults = (e) => {
    setSearchResults(
      searchIssues(e.target.value, list)
    )
  }

  return [ searchResults, newSearchResults ]
};


function searchIssues(search, itemList){
  if(search === '' || itemList === [])
    return [];


  var rx = new RegExp(`([^"]*${search}[^"]*)`,'gi');
  const listFiltered = itemList.filter(value => filterResults(value, rx))
  const slicedList = listFiltered.slice(0, 5)
  return slicedList;
}

function filterResults(value, rx){
  return matchResult(value.title, rx) || matchResult(value.labels, rx) || matchResult(value.body, rx)
}

function matchResult(value, rx){
  if (Array.isArray(value)) {
    var result = value.filter(label => {
      return String(label.name).match(rx)
    })

    return result.length > 0 ? true : false
  }

  return String(value).match(rx)
}
