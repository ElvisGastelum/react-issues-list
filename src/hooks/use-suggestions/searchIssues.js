export function searchIssues(search, itemList) {
  if (search === '' || itemList === [])
    return [];


  var rx = new RegExp(`([^"]*${search}[^"]*)`, 'gi');
  const listFiltered = itemList.filter(item => filterResults(item, rx));
  const slicedList = listFiltered.slice(0, 5);
  return slicedList;
}
function filterResults({ title, labels, body }, rx) {
  return matchResult(title, rx) || matchResult(labels, rx) || matchResult(body, rx);
}
function matchResult(item, rx) {
  if (Array.isArray(item)) {
    var result = item.filter(({ name }) => {
      return String(name).match(rx);
    });

    return result.length > 0 ? true : false;
  }

  return String(item).match(rx);
}
