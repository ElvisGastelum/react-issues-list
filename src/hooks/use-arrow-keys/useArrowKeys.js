import { useState } from 'react';

import { key } from './key-code';

export const useArrowKeys = (initialState, fn) => {
  const [ index, setIndex ] = useState(initialState);

  const handleArrowKeys = (keyCode, length) => {
    length = length - 1;
    switch (keyCode) {
      case key.DOWN:
        if (index < length) {
          setIndex(index + 1);
          break;
        }
        setIndex(0);
        break;
      case key.UP:
        if (index > 0) {
          setIndex(index - 1)
          break;
        }
        setIndex(length);
        break;
      case key.ENTER:
        fn();
        break;
      default:
    }
  }

  return [ index, handleArrowKeys ]
}
