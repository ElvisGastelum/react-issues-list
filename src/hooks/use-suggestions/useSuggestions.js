import { useState } from 'react';

import { useGithubIssuesList } from '../use-github-issues-list';
import { useArrowKeys } from '../use-arrow-keys';
import { searchIssues } from './searchIssues';

export const useSuggestions = (initialState, mockList) => {
  const list = mockList || useGithubIssuesList([]);
  const [ searchResults, setSearchResults ] = useState(initialState);
  const [  index, handleArrowKeys ] = useArrowKeys(0, handleEnter);

  const newSearchResults = value => {
    setSearchResults(
      searchIssues(value, list)
    )
  }

  const setArrowKeys = (keyCode, index) => {
    if (index) {
      handleArrowKeys(keyCode, index);  
    } else {
      handleArrowKeys(keyCode, searchResults.length);
    }
  }

  function handleEnter(){
    const itemSelected = searchResults[index].html_url;

    window.open(itemSelected);
  }


  return [ searchResults, newSearchResults, index, setArrowKeys ]
};



