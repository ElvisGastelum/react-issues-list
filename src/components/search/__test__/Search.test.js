import React from 'react';

import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { act } from 'react-dom/test-utils';

import Search from '../Search';
import data from './mockData.json';
import { key } from '../../../hooks/use-arrow-keys/key-code';
import { searchIssues } from '../../../hooks/use-suggestions/searchIssues';

Enzyme.configure({
  adapter: new Adapter(),
})

describe('Test <Search /> component', () => {

  const initialMarkupHTML = '<div class="search-container"><div class="search-box"><input type="text" placeholder="Search" class="search-input"><div class="search-items-list"><ul></ul></div></div></div>';
  const markupHTMLWithGoodProp = '<div class="search-container"><div class="search-box"><input type="text" placeholder="Search" class="search-input"><div class="search-items-list"><ul><li class="search-item active"><a target="_blank" rel="noopener noreferrer" href="https://github.com/facebook/react/issues/19362">Scheduler callback is only checked for null value and therefore throws when callback is undefined.</a></li><li class="search-item "><a target="_blank" rel="noopener noreferrer" href="https://github.com/facebook/react/pull/19349">fix(reset): set value tracker to default values after form reset</a></li><li class="search-item "><a target="_blank" rel="noopener noreferrer" href="https://github.com/facebook/react/issues/19020">Feature request: have Context.Provider throw error if missing `value` prop</a></li><li class="search-item "><a target="_blank" rel="noopener noreferrer" href="https://github.com/facebook/react/issues/18945">DevTools: Improve browser extension iframe support</a></li></ul></div></div></div>';
  
  test('should be render', async () => {

    var searchResults = [];
    const newSearchResults = value => {
      searchResults = searchIssues(value, data);
    }
    var index = 0;
    const setArrowKeys = e => {
      console.log(e);
    }
    
    const wrapper = mount(
      <Search
        searchResults={searchResults}
        newSearchResults={newSearchResults}
        index={index}
        setArrowKeys={setArrowKeys}
      />
    );
    expect(wrapper.html()).toBe(initialMarkupHTML);
    expect(wrapper).toMatchSnapshot();
  });

  test('should be regenerate html when pass new props', async () => {

    var searchResults = [];
    const newSearchResults = value => {
      searchResults = searchIssues(value, data);
    }
    var index = 0;
    const setArrowKeys = e => {
      console.log(e);
    }
      
    const wrapper = mount(
      <Search
        searchResults={searchResults}
        newSearchResults={newSearchResults}
        index={index}
        setArrowKeys={setArrowKeys}
      />
    );
      
    expect(wrapper.html()).toBe(initialMarkupHTML);
  
    act(() => {
      wrapper.find('input').simulate('change', { target: { value: 'good' } });
    })
  
    act(() => {
      wrapper.setProps({
        searchResults
      })
    })
  
    expect(wrapper.html()).toBe(markupHTMLWithGoodProp);
    expect(wrapper).toMatchSnapshot();
  });

  test('should be have a length 5 when search "react" label in the input', async () => {

    var searchResults = [];
    const newSearchResults = value => {
      searchResults = searchIssues(value, data);
    }
    var index = 0;
    const setArrowKeys = e => {
      console.log(e);
    }
      
    const wrapper = mount(
      <Search
        searchResults={searchResults}
        newSearchResults={newSearchResults}
        index={index}
        setArrowKeys={setArrowKeys}
      />
    );
      
    expect(wrapper.html()).toBe(initialMarkupHTML);
  
    act(() => {
      wrapper.find('input').simulate('change', { target: { value: 'react' } });
    })
  
    act(() => {
      wrapper.setProps({
        searchResults
      })
    })

    expect(wrapper.find('ul').children().length).toBe(5);
    expect(wrapper).toMatchSnapshot();
  });

  test('should be have a length 3 when search "good first issue" label in the input', async () => {

    var searchResults = [];
    const newSearchResults = value => {
      searchResults = searchIssues(value, data);
    }
    var index = 0;
    const setArrowKeys = e => {
      console.log(e);
    }
      
    const wrapper = mount(
      <Search
        searchResults={searchResults}
        newSearchResults={newSearchResults}
        index={index}
        setArrowKeys={setArrowKeys}
      />
    );
      
    expect(wrapper.html()).toBe(initialMarkupHTML);
  
    act(() => {
      wrapper.find('input').simulate('change', { target: { value: 'good first issue' } });
    })
  
    act(() => {
      wrapper.setProps({
        searchResults
      })
    })
    expect(wrapper.find('ul').children().length).toBe(3);
    expect(wrapper).toMatchSnapshot();
  });

  test('should be have active class the first suggestion item', async () => {

    var searchResults = [];
    const newSearchResults = value => {
      searchResults = searchIssues(value, data);
    }
    var index = 0;
    const setArrowKeys = e => {
      console.log(e);
    }
      
    const wrapper = mount(
      <Search
        searchResults={searchResults}
        newSearchResults={newSearchResults}
        index={index}
        setArrowKeys={setArrowKeys}
      />
    );
      
    expect(wrapper.html()).toBe(initialMarkupHTML);
  
    act(() => {
      wrapper.find('input').simulate('change', { target: { value: 'good first issue' } });
    })
  
    act(() => {
      wrapper.setProps({
        searchResults
      })
    })
    expect(wrapper.find('li').first().hasClass('active')).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  test('should be have active class the second suggestion item when press key.DOWN arrow', async () => {
    
    var index = 0;
    var searchResults = [];

    const newSearchResults = value => {
      searchResults = searchIssues(value, data);
    }
    const handleArrowKeys = (keyCode, length) => {
      length = length - 1;
      switch (keyCode) {
        case key.DOWN:
          console.log('key.DOWN press');
          if (index < length) {
            index++;
            break;
          }
          index = 0;
          break;
        case key.UP:
          if (index > 0) {
            index--;
            break;
          }
          index = length;
          break;
        case key.ENTER:
          console.log('Enter pressed');
          break;
        case key.MOUSE_EVENT:
          index = length;
          break;
        default:
          index = 0;
      }

    }
    const setArrowKeys = keyCode => {
      handleArrowKeys(keyCode, searchResults.length);
    }
      
    const wrapper = mount(
      <Search
        searchResults={searchResults}
        newSearchResults={newSearchResults}
        index={index}
        setArrowKeys={setArrowKeys}
      />
    );
      
    expect(wrapper.html()).toBe(initialMarkupHTML);
  
    act(() => {
      wrapper.find('input').simulate('change', { target: { value: 'good first issue' } });
    })
  
    act(() => {
      wrapper.setProps({
        searchResults
      })
    })
    
    expect(wrapper.find('li').first().hasClass('active')).toBe(true);
    
    act(() => {
      wrapper.find('input').simulate('keydown', {keyCode: key.DOWN})
    })
  
    act(() => {
      wrapper.setProps({
        index
      })
    })
    expect(wrapper.find('li').first().hasClass('active')).toBe(false);
    expect(wrapper.find('ul').childAt(1).hasClass('active')).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  test('should be have active class the last suggestion item when press key.UP arrow', async () => {
    
    var index = 0;
    var searchResults = [];

    const newSearchResults = value => {
      searchResults = searchIssues(value, data);
    }
    const handleArrowKeys = (keyCode, length) => {
      length = length - 1;
      switch (keyCode) {
        case key.DOWN:
          if (index < length) {
            index++;
            break;
          }
          index = 0;
          break;
        case key.UP:
          if (index > 0) {
            index--;
            break;
          }
          index = length;
          break;
        case key.ENTER:
          console.log('Enter pressed');
          break;
        case key.MOUSE_EVENT:
          index = length;
          break;
        default:
          index = 0;
      }

    }
    const setArrowKeys = keyCode => {
      handleArrowKeys(keyCode, searchResults.length);
    }
      
    const wrapper = mount(
      <Search
        searchResults={searchResults}
        newSearchResults={newSearchResults}
        index={index}
        setArrowKeys={setArrowKeys}
      />
    );
      
    expect(wrapper.html()).toBe(initialMarkupHTML);
  
    act(() => {
      wrapper.find('input').simulate('change', { target: { value: 'good first issue' } });
    })
  
    act(() => {
      wrapper.setProps({
        searchResults
      })
    })
    
    expect(wrapper.find('li').first().hasClass('active')).toBe(true);
    
    act(() => {
      wrapper.find('input').simulate('keydown', {keyCode: key.UP})
    })
  
    act(() => {
      wrapper.setProps({
        index
      })
    })
    expect(wrapper.find('li').first().hasClass('active')).toBe(false);
    expect(wrapper.find('li').last().hasClass('active')).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });



})

