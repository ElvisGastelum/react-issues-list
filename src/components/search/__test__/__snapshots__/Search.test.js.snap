// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Search /> component should be have a length 3 when search "good first issue" label in the input 1`] = `ReactWrapper {}`;

exports[`Test <Search /> component should be have a length 5 when search "react" label in the input 1`] = `ReactWrapper {}`;

exports[`Test <Search /> component should be have active class the first suggestion item 1`] = `ReactWrapper {}`;

exports[`Test <Search /> component should be have active class the last suggestion item when press key.UP arrow 1`] = `ReactWrapper {}`;

exports[`Test <Search /> component should be have active class the second suggestion item when press key.DOWN arrow 1`] = `ReactWrapper {}`;

exports[`Test <Search /> component should be regenerate html when pass new props 1`] = `ReactWrapper {}`;

exports[`Test <Search /> component should be render 1`] = `ReactWrapper {}`;
